src/nupic/algorithms/SpatialPooler.cpp

```diff
@@ -814,17 +823,17 @@ Real SpatialPooler::avgConnectedSpanForColumnND_(UInt column) const {
 }
 
 
-void SpatialPooler::adaptSynapses_(const UInt inputVector[], //TODO make sparse
-                                   const vector<UInt> &activeColumns) {
+void SpatialPooler::adaptSynapses_(SDR &input,
+                                   SDR &active) {
   vector<Real> permChanges(numInputs_, -1 * synPermInactiveDec_);
+  const auto &inputVector = input.getDense();
   for (UInt i = 0; i < numInputs_; i++) {
     if (inputVector[i] > 0) {
       permChanges[i] = synPermActiveInc_;
     }
   }
 
-  for (Size i = 0; i < activeColumns.size(); i++) {
-    const UInt column = activeColumns[i];
+  for(const auto &column : active.getFlatSparse()) {
     const vector<UInt> potential = potentialPools_.getSparseRow(column);
     vector<Real> perm(numInputs_, 0);
     permanences_.getRowToDense(column, perm);
```

Comment:
consider SDR perm; 
getRowToSparse()...

Reply:
I tried this and fail.  This is far bigger task than I thought because `adaptSynapses_` calls `updatePermanencesForColumn_` which expects its data in the dense format.  updatePermanencesForColumn_ is called in several places and it in turn calls several other methods.  I think that we should investigate doing this task as part of changing the SP from sparse matrix to connections backend, if we do that.

BTW: SparseMatrix.hpp is 11,000 lines and has 100 lines of unit tests (which are all commented out), yikes!

Reply:
I'll try to give it a go too, and if we fail, we'll leave it to other revamp. 

> BTW: SparseMatrix.hpp is 11,000 lines and has 100 lines of unit tests (which are all commented out), yikes! 

ouch!! I was suprised too, that SparseM is not a base for SparseBinaryM. Hope we can get rid off that class